File: 114-flatten-binary-tree-to-linked-list/flatten-binary-tree-to-linked-list.java
/**
 * Definition for a binary tree node.
 * public class TreeNode {
 *     int val;
 *     TreeNode left;
 *     TreeNode right;
 *     TreeNode() {}
 *     TreeNode(int val) { this.val = val; }
 *     TreeNode(int val, TreeNode left, TreeNode right) {
 *         this.val = val;
 *         this.left = left;
 *         this.right = right;
 *     }
 * }
 */
class Solution {
    public void flatten(TreeNode root) {
        Stack<TreeNode> s=new Stack<>();
        TreeNode ptr=root;
        if(root==null) return ;
        if(root.right!=null) s.push(root.right);

        while(ptr!=null || !s.isEmpty()){
            while(ptr.left!=null){
                ptr.right=ptr.left;
                ptr.left=null;
                ptr=ptr.right;
                if(ptr.right!=null)
                   s.push(ptr.right);
            }
            if(!s.isEmpty()){
                TreeNode temp=s.pop();
                ptr.right=temp;
                ptr.left=null;
                if(temp.right!=null)
                  s.push(temp.right);
                ptr=ptr.right;
            }
            else break;
        }

        return ;
    }
}

File: 114-flatten-binary-tree-to-linked-list/flatten-binary-tree-to-linked-list.py
# Definition for a binary tree node.
# class TreeNode(object):
#     def __init__(self, val=0, left=None, right=None):
#         self.val = val
#         self.left = left
#         self.right = right
class Solution(object):
    def flatten(self, root):
        """
        :type root: TreeNode
        :rtype: None Do not return anything, modify root in-place instead.
        """

        def check(node, pre):
            if node == None:
                return 
            
            pre[0].left = None
            pre[0].right = node
            pre[0] = node

            curRight = node.right
            check(node.left, pre)
            check(curRight, pre)
        

        dummy = TreeNode(-1)
        pre = [dummy]
        check(root, pre)

        return dummy.right